File: src/ffi/grovedb/src/element.cpp
// Copyright (c) 2026-present, The Dash Core developers
// Distributed under the MIT software license, see the accompanying
// file LICENSE.MIT or https://opensource.org/license/mit

#include <db_internal.h>
#include <util/ffi.h>

#include <grovedb/element.h>

#include <rust/grovedb_cxx/lib.h>

namespace grovedb {
Result<Element, Error> Element::Item(const Bytes& value)
{
  return CallFFI([&]() -> Result<Element, Error> {
    auto bytes = grovedb_cxx::grovedb_element_item(ToSlice(value));
    return Element::FromData(Bytes{bytes.begin(), bytes.end()});
  });
}

Result<Element, Error> Element::EmptyTree()
{
  return CallFFI([&]() -> Result<Element, Error> {
    auto bytes = grovedb_cxx::grovedb_element_empty_tree();
    return Element::FromData(Bytes{bytes.begin(), bytes.end()});
  });
}

Result<Element, Error> Element::EmptySumTree()
{
  return CallFFI([&]() -> Result<Element, Error> {
    auto bytes = grovedb_cxx::grovedb_element_empty_sum_tree();
    return Element::FromData(Bytes{bytes.begin(), bytes.end()});
  });
}

Result<Element, Error> Element::SumItem(int64_t value)
{
  return CallFFI([&]() -> Result<Element, Error> {
    auto bytes = grovedb_cxx::grovedb_element_sum_item(value);
    return Element::FromData(Bytes{bytes.begin(), bytes.end()});
  });
}
} // namespace grovedb

File: rust/grovedb_cxx/src/element.rs
//
// Copyright (c) 2026-present, The Dash Core developers
// SPDX-License-Identifier: MIT
// See the accompanying file LICENSE or https://opensource.org/license/mit
//

use grovedb_version::version::GroveVersion;

/// Serialize an Element to its bincode wire format.
pub(crate) fn serialize_element(
  element: &grovedb::Element,
  version: &GroveVersion,
) -> Result<Vec<u8>, String> {
  element
    .serialize(version)
    .map_err(|e| crate::ffi_error(grovedb::Error::from(e)))
}

/// Deserialize an Element from its bincode wire format.
pub(crate) fn deserialize_element(
  bytes: &[u8],
  version: &GroveVersion,
) -> Result<grovedb::Element, String> {
  grovedb::Element::deserialize(bytes, version)
    .map_err(|e| crate::ffi_error(grovedb::Error::from(e)))
}

// ---------------------------------------------------------------------------
// Element factory bridge functions
// ---------------------------------------------------------------------------

/// Create a serialized `Element::new_item` from raw value bytes.
pub fn grovedb_element_item(value: &[u8]) -> Result<Vec<u8>, String> {
  let version = GroveVersion::latest();
  let element = grovedb::Element::new_item(value.to_vec());
  serialize_element(&element, version)
}

/// Create a serialized `Element::empty_tree`.
pub fn grovedb_element_empty_tree() -> Result<Vec<u8>, String> {
  let version = GroveVersion::latest();
  let element = grovedb::Element::empty_tree();
  serialize_element(&element, version)
}

/// Create a serialized `Element::empty_sum_tree`.
pub fn grovedb_element_empty_sum_tree() -> Result<Vec<u8>, String> {
  let version = GroveVersion::latest();
  let element = grovedb::Element::empty_sum_tree();
  serialize_element(&element, version)
}

/// Create a serialized `Element::new_sum_item(value)`.
pub fn grovedb_element_sum_item(value: i64) -> Result<Vec<u8>, String> {
  let version = GroveVersion::latest();
  let element = grovedb::Element::new_sum_item(value);
  serialize_element(&element, version)
}
